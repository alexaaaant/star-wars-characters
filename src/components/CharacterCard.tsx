import { Card, CardContent, Grid, Typography } from '@mui/material';
import { observer } from 'mobx-react-lite';
import React from 'react';
import { Link } from 'react-router-dom';
import { Character } from '../types/characters';

interface CharacterCardProps {
  character: Character;
  observerRef: ((el: HTMLDivElement) => void) | null;
}

export const CharacterCard: React.FC<CharacterCardProps> = observer(
  ({ character, observerRef }) => {

    return (
      <>
        {character ? (
          <Grid item={true} xs={12} sm={6} md={4} key={character.name}>
            <Link
              to={`/character/${character.name}`}
              style={{ textDecoration: 'none' }}
            >
              <Card>
                <CardContent>
                  <Typography variant="h5" component="h2">
                    {character.name}
                  </Typography>
                </CardContent>
              </Card>
            </Link>
            {observerRef && <div ref={observerRef} />}
          </Grid>
        ) : (
          ''
        )}
      </>
    );
  }
);

CharacterCard.displayName = 'CharacterCard';
