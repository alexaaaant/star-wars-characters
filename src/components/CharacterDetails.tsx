export function CharacterDetails() {
  return <div>Details</div>;
}
