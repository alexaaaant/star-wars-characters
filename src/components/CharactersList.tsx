import { useState, useEffect, useCallback, useRef, ReactNode } from 'react';
import {
  CircularProgress,
  Container,
  Grid,
  TextField,
  Typography,
} from '@mui/material';
import { useDebounce } from '../common/hooks/useDebounce';
import { observer } from 'mobx-react-lite';
import { charactersStore } from '../stores/CharactersStore';
import { CharacterCard } from './CharacterCard';

export const CharactersList = observer(() => {
  const [searchQuery, setSearchQuery] = useState('');
  const debouncedSearchQuery = useDebounce(searchQuery);
  const observerComp = useRef<IntersectionObserver | null>(null);
  const initial = useRef(true);
  const { characters, startSearch, loadNextPage, isLoading } =
    charactersStore;

  useEffect(() => {
    if (initial.current && characters.size) {
      return;
    }
    initial.current = false;
    startSearch(debouncedSearchQuery);
  },        [debouncedSearchQuery, startSearch]);

  const observerRef = useCallback(
    (node: HTMLDivElement) => {
      if (!observerComp.current) {
        observerComp.current = new IntersectionObserver((entries) => {
          if (entries[0].isIntersecting) {
            loadNextPage();
          }
        });
      }

      if (node !== null) {
        observerComp.current.observe(node);
      } else {
        observerComp.current.disconnect();
      }
    },
    [loadNextPage]
  );

  const handleSearchChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(event.target.value);
  };

  const charactersElements = () => {
    const elements: ReactNode[] = [];
    let i = 0;
    characters.forEach((character) => {
      elements.push(
        <CharacterCard
          key={character.name}
          character={character}
          observerRef={i === characters.size - 1 ? observerRef : null}
        />
      );
      i++;
    });
    return elements;
  };

  return (
    <Container>
      <TextField
        label="Search"
        value={searchQuery}
        onChange={handleSearchChange}
        sx={{ m: 2 }}
      />
      {!!characters.size && (
        <Grid container={true} spacing={2}>
          {charactersElements()}
        </Grid>
      )}
      <Grid container={true} justifyContent={'center'}>
        {!characters.size && !isLoading && (
          <Typography variant="h5" component="h2">
            Empty
          </Typography>
        )}
        {isLoading && <CircularProgress sx={{ m: 2 }} />}
      </Grid>
    </Container>
  );
});

CharactersList.displayName = 'CharactersList';
