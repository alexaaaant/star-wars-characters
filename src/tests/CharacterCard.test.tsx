import { render, screen } from '@testing-library/react';
import { CharacterCard } from '../components/CharacterCard';
import '@testing-library/jest-dom';

jest.mock('react-router-dom', () => ({
  Link: ({ to, children }: any) => <a href={to}>{children}</a>,
}));

describe('CharacterCard', () => {
  const character = {
    name: 'Luke Skywalker',
    birth_year: '19BBY',
    gender: 'n/a',
    height: '300',
  };

  it('makes a snapshot', () => {
    const { container } = render(
      <CharacterCard character={character} observerRef={null} />
    );

    expect(container).toMatchSnapshot();
  });

  it('renders a link to character details', () => {
    render(<CharacterCard character={character} observerRef={null} />);

    const linkElement = screen.getByRole('link', { name: `${character.name} ${character.birth_year}` });

    expect(linkElement).toHaveAttribute('href', `/character/${character.name}`);
  });

  it('renders observer div when observerRef is provided', () => {
    const observerRefMock = jest.fn();
    render(
      <CharacterCard character={character} observerRef={observerRefMock} />
    );

    const observerDiv = screen.getByTestId('observer-div');

    expect(observerDiv).toBeInTheDocument();
    expect(observerRefMock).toHaveBeenCalledTimes(1);
  });

  it('does not render observer div when observerRef is null', () => {
    render(<CharacterCard character={character} observerRef={null} />);

    const observerDiv = screen.queryByTestId('observer-div');

    expect(observerDiv).not.toBeInTheDocument();
  });
});
