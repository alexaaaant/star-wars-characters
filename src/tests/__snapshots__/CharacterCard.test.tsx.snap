// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard makes a snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 css-13i4rnv-MuiGrid-root"
  >
    <a
      href="/character/Luke Skywalker"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
          >
            Luke Skywalker
          </h2>
          <h6
            class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
          >
            19BBY
          </h6>
        </div>
      </div>
    </a>
  </div>
</div>
`;
